ref-exchange/src/account_deposit.rs

```diff
@@ -25,74 +26,130 @@ const U32_STORAGE: StorageUsage = 4;
 const ACC_ID_AS_KEY_STORAGE: StorageUsage = ACC_ID_STORAGE + 4;
+const KEY_PREFIX_ACC: StorageUsage = 64;
 /// As a near_sdk::collection key, 1 byte for prefiex
 const ACC_ID_AS_CLT_KEY_STORAGE: StorageUsage = ACC_ID_AS_KEY_STORAGE + 1;
 
 // ACC_ID: the Contract accounts map key length
 // + VAccount enum: 1 byte
 // + U128_STORAGE: near_amount storage
+// + U32_STORAGE: legacy_tokens HashMap length
 // + U32_STORAGE: tokens HashMap length
 // + U64_STORAGE: storage_used
 pub const INIT_ACCOUNT_STORAGE: StorageUsage =
-    ACC_ID_AS_CLT_KEY_STORAGE + 1 + U128_STORAGE + U32_STORAGE + U64_STORAGE;
+    ACC_ID_AS_CLT_KEY_STORAGE + 1 + U128_STORAGE + U32_STORAGE + U32_STORAGE + U64_STORAGE;
 
 #[derive(BorshDeserialize, BorshSerialize)]
 pub enum VAccount {
+    V1(AccountV1),
     Current(Account),
 }
 
-impl From<Account> for VAccount {
-    fn from(account: Account) -> Self {
-        VAccount::Current(account)
+impl VAccount {
+    /// Upgrades from other versions to the currently used version.
+    pub fn into_current(self, account_id: &AccountId) -> Account {
+        match self {
+            VAccount::Current(account) => account,
+            VAccount::V1(account) => account.into_current(account_id),
+        }
     }
 }
 
-impl From<VAccount> for Account {
-    fn from(v_account: VAccount) -> Self {
-        match v_account {
-            VAccount::Current(account) => {account},
-        }
+impl From<Account> for VAccount {
+    fn from(account: Account) -> Self {
+        VAccount::Current(account)
     }
 }
 
 
 /// Account deposits information and storage cost.
-#[derive(BorshSerialize, BorshDeserialize, Default, Clone)]
+#[derive(BorshSerialize, BorshDeserialize)]
 pub struct Account {
     /// Native NEAR amount sent to the exchange.
     /// Used for storage right now, but in future can be used for trading as well.
     pub near_amount: Balance,
     /// Amounts of various tokens deposited to this account.
-    pub tokens: HashMap<AccountId, Balance>,
+    pub legacy_tokens: HashMap<AccountId, Balance>,
+    pub tokens: UnorderedMap<AccountId, Balance>,
     pub storage_used: StorageUsage,
 }
 
+impl Account {
+    pub fn new(account_id: &AccountId) -> Self {
+        Account {
+            near_amount: 0,
+            legacy_tokens: HashMap::new(),
+            tokens: UnorderedMap::new(StorageKey::AccountTokens {
+                account_id: account_id.clone(),
+            }),
+            storage_used: 0,
+        }
+    }
 
+    /// Deposit amount to the balance of given token,
+    /// if given token not register and not enough storage, deposit fails 
+    pub(crate) fn deposit_with_storage_check(&mut self, token: &AccountId, amount: Balance) -> bool { 
+        if let Some(balance) = self.tokens.get(token) {
+            // token has been registered, just add without storage check, 
+            let new_balance = balance
+                .checked_add(amount)
+                .expect("errors::BALANCE_OVERFLOW");
```

Comment:
Need to consistently use `checked_add` or not use it at all because `overflow_checks = true`